lib/utils.js

```diff
@@ -61,22 +61,38 @@ function parseCapsForInnerDriver (jsonwpCapabilities, w3cCapabilities, constrain
 
   const {W3C, MJSONWP} = BaseDriver.DRIVER_PROTOCOL;
 
-  // Make copies of the capabilities that include the default capabilities
-  if (hasW3CCaps) {
-    w3cCapabilities = {
-      ...w3cCapabilities,
-      alwaysMatch: {
-        ...defaultCapabilities,
-        ...w3cCapabilities.alwaysMatch,
-      },
-    };
-  }
+  // Make sure we don't mutate the original arguments
+  jsonwpCapabilities = _.cloneDeep(jsonwpCapabilities);
+  w3cCapabilities = _.cloneDeep(w3cCapabilities);
+  defaultCapabilities = _.cloneDeep(defaultCapabilities);
 
-  if (hasJSONWPCaps) {
-    jsonwpCapabilities = {
-      ...defaultCapabilities,
-      ...jsonwpCapabilities,
-    };
+  if (!_.isEmpty(defaultCapabilities)) {
+    if (hasW3CCaps) {
+      const {firstMatch = [], alwaysMatch = {}} = w3cCapabilities;
+      for (const [defaultCapKey, defaultCapValue] of _.toPairs(defaultCapabilities)) {
+        let isCapAlreadySet = false;
+        for (const firstMatchEntry of firstMatch) {
+          if (_.has(removeW3CPrefixes(firstMatchEntry), removeW3CPrefix(defaultCapKey))) {
+            isCapAlreadySet = true;
+            break;
+          }
+        }
+        isCapAlreadySet = isCapAlreadySet || _.has(removeW3CPrefixes(alwaysMatch), removeW3CPrefix(defaultCapKey));
+        if (isCapAlreadySet) {
+          continue;
+        }
+
+        // Only add the default capability if it is not overridden
+        if (_.isEmpty(firstMatch)) {
+          w3cCapabilities.firstMatch = [{defaultCapKey: defaultCapValue}];
```

Comment:
doesn't this need to be `[{[defaultCapKey]: defaultCapValue}]`?

Reply:
good catch, will fix